databricks_cli/tunnel/server.py

```diff
@@ -0,0 +1,164 @@
+# Databricks CLI
+# Copyright 2017 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"), except
+# that the use of services to which certain application programming
+# interfaces (each, an "API") connect requires that the user first obtain
+# a license for the use of the APIs from Databricks, Inc. ("Databricks"),
+# by creating an account at www.databricks.com and agreeing to either (a)
+# the Community Edition Terms of Service, (b) the Databricks Terms of
+# Service, or (c) another written agreement between Licensee and Databricks
+# for the use of the APIs.
+#
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import asyncio
+
+from urllib.parse import urljoin
+
+import aiohttp
+import socketio
+
+from aiohttp import BasicAuth
+from socketio.exceptions import BadNamespaceError
+from tornado.iostream import IOStream, StreamClosedError
+from tornado.tcpserver import TCPServer
+
+
+DBX_TUNNEL_SERVER_PORT = 6006
+
+
+logger = logging.getLogger(__name__)
+
+
+class TunnelConfig:
+    def __init__(self, databricks_config, org_id, cluster_id):
+        self.host = databricks_config.host
+        self.token = databricks_config.token
+        self.org_id = org_id
+        self.cluster_id = cluster_id
+        self.remote_port = DBX_TUNNEL_SERVER_PORT
+        # TODO(tunneling-cli): this assumes a multitenant workspace, we will need to confirm if
+        #  this is still the case for st workspace
+        self.endpoint_path = \
+            f"driver-proxy-api/o/{self.org_id}/{self.cluster_id}/{self.remote_port}/"
+
+    @property
+    def endpoint_url(self):
+        return urljoin(self.host, self.endpoint_path)
+
+    @property
+    def healthcheck_url(self):
+        return f"{self.endpoint_url}health"
+
+    @property
+    def socket_path(self):
+        return f"{self.endpoint_path}socket.io"
+
+
+class TunnelClient:
+    def __init__(self, stream: IOStream, session: aiohttp.ClientSession,
+                 tunnel_config: TunnelConfig, debug=False):
+        self.stream = stream
+        self.tunnel_config = tunnel_config
+        self.io = socketio.AsyncClient(http_session=session, logger=debug, engineio_logger=debug)
```

Comment:
nit: OK to start with `debug` logging, but let's have a TODO to remove this

Reply:
i think this is important for debugging problems. currently it receives a boolean flag `--debug` from the cli. if it's true then the tunneling loggings will be shown. but currently, `--debug` flag is shared with other cli cmds. i wonder if we should surface new debug flag for tunneling

Reply:
If there is a `--debug` flag that toggles the debug logging off/on, that's actually perfect.